src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ExternalTypeMapEntryNode.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    internal sealed class ExternalTypeMapEntryNode(TypeDesc typeMapGroup, string key, TypeDesc targetType) : DependencyNodeCore<NodeFactory>, ISortableNode
+    {
+        public override bool InterestingForDynamicDependencyAnalysis => false;
+
+        public override bool HasDynamicDependencies => false;
+
+        public override bool HasConditionalStaticDependencies => false;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        public TypeDesc TypeMapGroup { get; } = typeMapGroup;
+
+        public string Key { get; } = key;
+
+        public TypeDesc TargetType { get; } = targetType;
+
+        public int ClassCode => -785190502;
+
+        public int CompareToImpl(ISortableNode other, CompilerComparer comparer)
+        {
+            ExternalTypeMapEntryNode otherEntry = (ExternalTypeMapEntryNode)other;
+            if (TypeMapGroup != otherEntry.TypeMapGroup)
+            {
+                return comparer.Compare(TypeMapGroup, otherEntry.TypeMapGroup);
+            }
+            else if (Key != otherEntry.Key)
+            {
+                return string.Compare(Key, otherEntry.Key, StringComparison.Ordinal);
+            }
+            else if (TargetType != otherEntry.TargetType)
+            {
+                return comparer.Compare(TargetType, otherEntry.TargetType);
+            }
+
+            return 0;
+        }
+
+        public override IEnumerable<CombinedDependencyListEntry> GetConditionalStaticDependencies(NodeFactory context) => Array.Empty<CombinedDependencyListEntry>();
+
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory context)
+        {
+            return [
+                new DependencyListEntry(context.MaximallyConstructableType(TargetType), "Target type in external type map")
```

Comment:
Is this for the value in an external type map entry? Why is it marked as maximally constructible? My understanding is that it would be minimally constructible (i.e. just get me the `Type` object with nothing else, just enough so I can look up attributes). Then if any attribute has something like `new TheType()` in an instance method, only then would the type actually be marked as constructible.

Eg. in CsWinRT we have external type map entries for all generic instantiations, but those would point to proxy types, which we never need to instantiate at all. They're only needed to lookup their attribute which then contains the actual marshalling logic we need to invoke.

Am I understanding this wrong? Or is this something we could potentially improve?

Reply:
A `MaximallyConstructableType` is the "you can get custom attributes, allocate instances call a method" node. A `NecessaryType` won't let you get a System.Type instance that's usable.

Reply:
Mmh am I confusing some stuff that Michal perhaps mentioned he _wanted_ to add but that is not available just yet then? Basically I thought there was a way to get a `Type` instance that you can use strictly only to get attributes from and nothing else. I remember we hit that scenario in several cases around IDIC, for instance. Does that ring any bells? 🤔